manim: remove old manim(lib) package and init at 0.15.2

mofrim opened the pull request:
###### Motivation for this change
There is an existing python application called `manim` in nixpkgs, but, it is first of all incorrectly named (on PyPi it is called manimlib) and it's development under *this* name is discontinued. The original `manim`-package was created and is still being developed by the author of the Youtube-math-channel 3blue1brown under the new name `manimgl`. The version I packaged here is a community-developed fork, which is supposed to be more stable and clean. It brings along 4 dependencies `mapbox-earcut`, `cloup`, `isosurfaces` and `srt`.

In order to make this package the new `manim` package in nixpkgs, i did remove the old manim(lib) package in favor of the new one.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

mofrim committed:
python3Packages.mapbox-earcut: init at 2.2.2

fufexan commented:
Result of `nixpkgs-review pr 120479` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>manim-community</li>
    <li>python38Packages.manimpango</li>
    <li>python38Packages.mapbox-earcut</li>
    <li>python39Packages.manimpango</li>
    <li>python39Packages.mapbox-earcut</li>
  </ul>
</details>

SuperSandro2000 commented on pkgs/development/python-modules/mapbox-earcut/default.nix:
```diff
@@ -0,0 +1,42 @@
+        polygons.
+      '';
+      maintainers = with maintainers; [ friedelino ];
+      broken = pybind11.version < "2.6.0"; # broken because of missing submodules in lower versions
```
```suggestion
```

We have version 2.6.1 so we don't need this. https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/pybind11/default.nix#L17

SuperSandro2000 commented on pkgs/development/python-modules/mapbox-earcut/default.nix:
```diff
@@ -0,0 +1,42 @@
+    buildInputs = [ pybind11 ];
+    propagatedBuildInputs = [ numpy ];
+
+    # Some test failed, some passed...
+    checkInputs = [ pytestCheckHook ];
+    doCheck = false;
```
How many failed? Can we disable them? Why do they fail?

friedelino replied:
Output of pytestCheckPhase:
```
============================= test session starts ==============================
platform linux -- Python 3.8.8, pytest-6.2.2, py-1.9.0, pluggy-0.13.1
rootdir: /build/source
collected 13 items

tests/test_earcut.py ......FFFFFF.                                       [100%]

=================================== FAILURES ===================================
___________________________ test_end_index_too_large ___________________________

    def test_end_index_too_large():
        verts = np.array([[0, 0], [1, 0], [1, 1]]).reshape(-1, 2)
        rings = np.array([5])

>       with pytest.raises(ValueError, message="Expecting ValueError"):
E       TypeError: Unexpected keyword arguments passed to pytest.raises: message
E       Use context-manager form instead?

tests/test_earcut.py:77: TypeError
___________________________ test_end_index_too_small ___________________________

    def test_end_index_too_small():
        verts = np.array([[0, 0], [1, 0], [1, 1]]).reshape(-1, 2)
        rings = np.array([2])

>       with pytest.raises(ValueError, message="Expecting ValueError"):
E       TypeError: Unexpected keyword arguments passed to pytest.raises: message
E       Use context-manager form instead?

tests/test_earcut.py:85: TypeError
______________________________ test_end_index_neg ______________________________

    def test_end_index_neg():
        verts = np.array([[0, 0], [1, 0], [1, 1]]).reshape(-1, 2)
        rings = np.array([-1])

>       with pytest.raises(ValueError, message="Expecting ValueError"):
E       TypeError: Unexpected keyword arguments passed to pytest.raises: message
E       Use context-manager form instead?

tests/test_earcut.py:93: TypeError
__________________________ test_rings_not_increasing ___________________________

    def test_rings_not_increasing():
        verts = np.array([[0, 0], [1, 0], [1, 1]]).reshape(-1, 2)
        rings = np.array([3, 0, 3])

>       with pytest.raises(ValueError, message="Expecting ValueError"):
E       TypeError: Unexpected keyword arguments passed to pytest.raises: message
E       Use context-manager form instead?

tests/test_earcut.py:101: TypeError
_______________________________ test_rings_same ________________________________

    def test_rings_same():
        verts = np.array([[0, 0], [1, 0], [1, 1]]).reshape(-1, 2)
        rings = np.array([3, 3])

>       with pytest.raises(ValueError, message="Expecting ValueError"):
E       TypeError: Unexpected keyword arguments passed to pytest.raises: message
E       Use context-manager form instead?

tests/test_earcut.py:109: TypeError
________________________________ test_no_rings _________________________________

    def test_no_rings():
        verts = np.array([[0, 0], [1, 0], [1, 1]]).reshape(-1, 2)
        rings = np.array([])

>       with pytest.raises(ValueError, message="Expecting ValueError"):
E       TypeError: Unexpected keyword arguments passed to pytest.raises: message
E       Use context-manager form instead?

tests/test_earcut.py:125: TypeError
=========================== short test summary info ============================
FAILED tests/test_earcut.py::test_end_index_too_large - TypeError: Unexpected...
FAILED tests/test_earcut.py::test_end_index_too_small - TypeError: Unexpected...
FAILED tests/test_earcut.py::test_end_index_neg - TypeError: Unexpected keywo...
FAILED tests/test_earcut.py::test_rings_not_increasing - TypeError: Unexpecte...
FAILED tests/test_earcut.py::test_rings_same - TypeError: Unexpected keyword ...
FAILED tests/test_earcut.py::test_no_rings - TypeError: Unexpected keyword ar...
========================= 6 failed, 7 passed in 0.20s ==========================
```
... looks like the tests don't fit the current pytest-syntax. Could fix that through patching the tests.

friedelino replied:
i think it is done

SuperSandro2000 commented on pkgs/development/python-modules/mapbox-earcut/default.nix:
```diff
@@ -0,0 +1,42 @@
+    sha256 = "14z2vj13a1qcaqc6jvg00f367la5470bclvfi9mmlvpac3hhhywd";
+    };
+
+    nativeBuildInputs = [ setuptools ];
```
Is that required at runtime?

friedelino replied:
No. But it is required at buildtime. So I thought putting in into `nativeBuildInputs` wouldn't be a bad idea. Isn't `nativeBuildInputs` meant for specifying buildtime deps? I know i don't need to specify this because `setuptoolsBuildPhase` will automatically use `setuptools`.

SuperSandro2000 commented on pkgs/development/python-modules/mapbox-earcut/default.nix:
```diff
@@ -0,0 +1,42 @@
+
+buildPythonPackage rec {
+  pname = "mapbox-earcut";
+  version = "2.2.2";
```
This version seems arbitrary.

friedelino replied:
You're right. Something went wrong about this. `0.12.10` would be right. Will change that.

SuperSandro2000 commented on pkgs/development/python-modules/mapbox-earcut/default.nix:
```diff
@@ -0,0 +1,42 @@
+  src = fetchFromGitHub {
+    owner = "skogler";
+    repo = "mapbox_earcut_python";
+    rev = "257c2c82819acf72b2b757f200c160e55b9427fc";
```
Why are we not fetching https://github.com/skogler/mapbox_earcut_python/releases/tag/v0.12.10?

friedelino replied:
Because this repo does not receive a lot of commits, but after v0.12.10 release there have been some substantial commits that i would not want to miss. Specifically switching from cmake to complete setuptools build-system.

SuperSandro2000 replied:
Then we need to change the version number to unstable-XXXX-XX-XX where the XXXX-XX-XX is the date of the commit.

friedelino replied:
ah alright! fixed.

SuperSandro2000 commented on pkgs/development/python-modules/manimpango/default.nix:
```diff
@@ -0,0 +1,85 @@
+      used in Manim to render (non-LaTeX) text.
+    '';
+    homepage = "https://github.com/ManimCommunity/ManimPango";
+    license = licenses.gpl3;
```
gpl3Plus or gpl3Only?

friedelino replied:
by the github license page gpl3only

SuperSandro2000 commented on pkgs/development/python-modules/manimpango/default.nix:
```diff
@@ -0,0 +1,85 @@
+
+  # checks fail until manim is installed
+  doCheck = false;
+  checkInputs = [ pytest ];
+  checkPhase = ''
+    pytest
+  '';
```
```suggestion
  checkInputs = [ pytestCheckHook ];
```

SuperSandro2000 commented on pkgs/development/python-modules/manimpango/default.nix:
```diff
@@ -0,0 +1,85 @@
+  checkPhase = ''
+    pytest
+  '';
+
+  disabled = pythonOlder "3.8";
```
```suggestion
```

SuperSandro2000 commented on pkgs/development/python-modules/manimpango/default.nix:
```diff
@@ -0,0 +1,85 @@
+
+buildPythonPackage rec {
+  pname = "manimpango";
+  version = "0.2.6";
```
```suggestion
  version = "0.2.6";
  disabled = pythonOlder "3.8";
```

SuperSandro2000 commented on pkgs/development/python-modules/manimpango/default.nix:
```diff
@@ -0,0 +1,85 @@
+  # means that characters should be converted to lowercase [..]
+  src = fetchPypi {
+    pname = "ManimPango";
+    version = "0.2.6";
```
```suggestion
    inherit version;
```

SuperSandro2000 commented on pkgs/development/python-modules/manimpango/default.nix:
```diff
@@ -0,0 +1,85 @@
+  # inputs only needed during build
+  nativeBuildInputs = [
+    pkg-config
+    setuptools
```
Is that required at runtime?

SuperSandro2000 commented on pkgs/development/python-modules/manimpango/default.nix:
```diff
@@ -0,0 +1,85 @@
+    cython
+  ];
+
+  # non-python deps
```
```suggestion
```

friedelino replied:
is this wrong?

SuperSandro2000 replied:
Those comments don't add any value to the package. We are in the pythonPackage part of nixpkgs and every package here is either a pythonPackage or a build/runtime dependency.

SuperSandro2000 commented on pkgs/development/python-modules/manimpango/default.nix:
```diff
@@ -0,0 +1,85 @@
+    pango
+  ];
+
+  # python deps
```
```suggestion
```

friedelino replied:
also wrong?

SuperSandro2000 commented on pkgs/applications/video/manim-community/default.nix:
```diff
@@ -0,0 +1,98 @@
+    mv $out/bin/manim $out/bin/manim-com
+    mv $out/bin/manimce $out/bin/manimce-com
+  '';
+
+  disabled = pythonOlder "3.8";
```
```suggestion
```

SuperSandro2000 commented on pkgs/applications/video/manim-community/default.nix:
```diff
@@ -0,0 +1,98 @@
+
+buildPythonApplication rec {
+  pname = "manim-community";
+  version = "0.5.0";
```
```suggestion
  version = "0.5.0";
  disabled = pythonOlder "3.8";
```

SuperSandro2000 commented on pkgs/applications/video/manim-community/default.nix:
```diff
@@ -0,0 +1,98 @@
+
+  patches = [ ./remove-dependency-constraints.patch ];
+
+  nativeBuildInputs =  [ poetry-core setuptools ];
```
Is setuptools required at runtime?

SuperSandro2000 commented on pkgs/applications/video/manim-community/default.nix:
```diff
@@ -0,0 +1,98 @@
+    sox
+    ffmpeg_3
+    xdg-utils
+    texlive.combined.scheme-basic   # not too happy about this.. but it is hard to
+                                    # declare which tex-packages are exactly
+                                    # needed... maybe TODO
```
```suggestion
    # TODO: figure out which packages are exactly required
    texlive.combined.scheme-basic
```

SuperSandro2000 commented on pkgs/applications/video/manim-community/default.nix:
```diff
@@ -0,0 +1,98 @@
+                                    # needed... maybe TODO
+  ];
+
+  doCheck = false;
+  checkInputs = [ pytestcov ];
+  checkPhase = ''
+    pytest
+  '';
```
```suggestion
  checkInputs = [ pytestCheckHook ];
```

Please remove the pytest-cov flags with substituteInPlace.

collares commented on pkgs/top-level/python-packages.nix:
```diff
@@ -4011,6 +4013,18 @@ in {
+
+  marionette_driver = callPackage ../development/python-modules/marionette-harness/marionette_driver.nix { };
+
+  marionette-harness = callPackage ../development/python-modules/marionette-harness { };
```
ofBorg is complaining about this line. Was this the result of a merge?

friedelino replied:
hmmm, yes there was a merge of master into this branch. but i don't understand where this comes from. I didn't touch marionette*

friedelino replied:
found out i messed up the merge. now it should be okay

collares commented on pkgs/applications/video/manim-community/default.nix:
```diff
@@ -0,0 +1,94 @@
+    sha256 = "e4284648cbb5d463c6dbced1d4e60fa9b68c14f4bb3b10c8eab0f8f49ca5da88";
+  };
+
+  patches = [ ./remove-dependency-constraints.patch ];
```
Maybe something like this instead?

```nix
        postPatch = ''
          # remove dependency constraints
          sed 's/>=[0-9.]\+,<[0-9.]\+//' -i setup.py
        '';
```

friedelino replied:
cool! will do it that way.

collares commented:
Unfortunately `moderngl` was updated yesterday and it's now broken because [glcontext](https://github.com/moderngl/glcontext) isn't packaged.

SuperSandro2000 commented on pkgs/applications/video/manim-community/default.nix:
```diff
@@ -0,0 +1,94 @@
+    texlive.combined.scheme-basic
+  ];
+
+  doCheck = false;
+  checkInputs = [ pytestCheckHook ];
```
Why are tests disabled?

friedelino replied:
See comment in current version of the file

SuperSandro2000 commented on pkgs/development/python-modules/manimpango/default.nix:
```diff
@@ -0,0 +1,79 @@
+
+  # checks fail until manim is installed
+  doCheck = false;
+  checkInputs = [ pytestCheckHook ];
```
```suggestion
```

friedelino replied:
fixed

friedelino commented:
> Unfortunately `moderngl` was updated yesterday and it's now broken because [glcontext](https://github.com/moderngl/glcontext) isn't packaged.
> 
> cc @c0deaddict

should be fixed if https://github.com/NixOS/nixpkgs/pull/121439 gets merged

collares commented:
@ofborg build

IvarWithoutBones commented on pkgs/applications/video/manim-community/default.nix:
```diff
@@ -0,0 +1,141 @@
+    mv $out/bin/manimce $out/bin/manimce-com
+  '';
+
+  meta = {
+    description = "Animation engine for explanatory math videos - Community version";
+    longDescription = ''
+      Manim is an animation engine for explanatory math videos. It's used to
+      create precise animations programmatically, as seen in the videos of
+      3Blue1Brown on YouTube. This is the community maintained version of
+      manim.
+    '';
+    homepage = "https://github.com/ManimCommunity/manim";
+    license = lib.licenses.mit;
+    maintainers = with lib.maintainers; [ friedelino ];
+  };
+}
```
```suggestion
  meta = with lib; {
    ...
    license = licenses.mit;
    maintainers = with maintainers; [ friedelino ];
  };
}
```

friedelino replied:
:+1:

IvarWithoutBones commented on pkgs/development/python-modules/manimpango/pangocairo.patch:
```diff
@@ -0,0 +1,21 @@
+diff --git a/setup.py b/setup.py
```
Perhaps this could be done using `substituteInPlace`, to avoid breakage in future updates?

friedelino replied:
done!

IvarWithoutBones commented on pkgs/development/python-modules/mapbox-earcut/tests.patch:
```diff
@@ -0,0 +1,67 @@
+diff --git a/tests/test_earcut.py b/tests/test_earcut.py
```
Same comment as above, especially since this is a repeating pattern.
```bash
substituteInPlace tests/test_earcut.py --replace \
   "pytest.raises(ValueError, message(\"Expecting ValueError\")" "with pytest.raises(ValueError)"

friedelino replied:
done!

IvarWithoutBones reviewed (changes requested):
A few nitpicks, but overal LGTM :)

SuperSandro2000 commented on pkgs/applications/video/manim-community/default.nix:
```diff
@@ -0,0 +1,141 @@
+
+    cairo
+    sox
+    ffmpeg_3
```
Please use ``ffmpeg`` (4).

friedelino replied:
Ok, done.

SuperSandro2000 commented on pkgs/development/python-modules/manimpango/default.nix:
```diff
@@ -0,0 +1,76 @@
+  # pangocairo is part of normal cairo-package, so remove corresponding deps
+  # from setup.py
+  postPatch =''
+    substituteInPlace setup.py --replace \
+    "\"pangocairo-1.0\"," ""
+
+    substituteInPlace setup.py --replace \
+    "_pkg_config = PKG_CONFIG(\"pangocairo\")" "_pkg_config = PKG_CONFIG(\"pango\")"
```
```suggestion
    substituteInPlace setup.py \
      --replace "\"pangocairo-1.0\"," "" \
      --replace "_pkg_config = PKG_CONFIG(\"pangocairo\")" "_pkg_config = PKG_CONFIG(\"pango\")"
```

collares commented:
It would be great to have this in 21.05!

dotlambda commented on pkgs/top-level/all-packages.nix:
```diff
@@ -23533,6 +23533,8 @@ in
     opencv = python3Packages.opencv3;
   };
 
+  manim-community = python3Packages.callPackage ../applications/video/manim-community { };
```
```suggestion
  manim-community = callPackage ../applications/video/manim-community { };
```

friedelino replied:
okay, will do that

dotlambda replied:
You'll have to use `python3.pkgs.buildPythonApplication`.
If use use `python3Packages.callPackage` overriding specific Python dependencies will be harder.

friedelino replied:
done

dotlambda commented on pkgs/development/python-modules/mapbox-earcut/default.nix:
```diff
@@ -0,0 +1,45 @@
+
+  checkInputs = [ pytestCheckHook ];
+
+  doCheck = true;
```
```suggestion
```
It's the default for Python packages.

friedelino replied:
sure, will do that

dotlambda commented on pkgs/development/python-modules/mapbox-earcut/default.nix:
```diff
@@ -0,0 +1,45 @@
+  };
+
+  postPatch =''
+    substituteInPlace tests/test_earcut.py --replace \
+    "pytest.raises(ValueError, message=\"Expecting ValueError\")" \
+    "pytest.raises(ValueError)"
```
```suggestion
    substituteInPlace tests/test_earcut.py --replace \
      "pytest.raises(ValueError, message=\"Expecting ValueError\")" \
      "pytest.raises(ValueError)"
```

friedelino commented on pkgs/applications/video/manim-community/default.nix:
```diff
@@ -0,0 +1,141 @@
+    (texlive.combine manim-tinytex)
+  ];
+
+  # no tests included with PyPi package. TODO: use github repo as source and
+  # include tests
```
honestly, i tried this. but most of the tests need display although it would be possible to write tests that work with nix, i really do not have time to do this. You or anybody else would be really welcome to help out on this.

mainrs replied:
Should we mark this as outdated and add tests later on? It's not that easy to be honest since, as friedelino mentioned, the repository uses poetry.

collares replied:
It might be just a matter of passing `format = "pyproject";` to buildPythonApplication, adding `poetry-core` to `nativeBuildInputs` and unpinning dependencies. I agree it should be done in a separate PR, though.

friedelino replied:
should be obsolete now

friedelino commented on pkgs/applications/video/manim-community/default.nix:
```diff
@@ -0,0 +1,141 @@
+, xdg-utils
+, texlive
+
+, python
+, colour
+, numpy
+, pillow
+, scipy
+, tqdm
+, pydub
+, pygments
+, rich
+, pycairo
+, manimpango
+, networkx
+, setuptools
+, importlib-metadata
+, grpcio
+, grpcio-tools
+, watchdog
+, jupyterlab
+, moderngl
+, moderngl-window
+, mapbox-earcut
```
okay, will do that

friedelino commented on pkgs/development/python-modules/mapbox-earcut/default.nix:
```diff
@@ -0,0 +1,45 @@
+    "pytest.raises(ValueError)"
+  '';
+
+  buildInputs = [ pybind11 ];
```
ok

IvarWithoutBones replied:
It doesn't look like this was adressed

Aathish04 commented:
@friedelino The two commands one can use to invoke ManimCE are `manim` and `manimce`. Instead of `manim-com` being the command for Nix, I think changing it to `manimce` would make things more consistenc.

naveen521kk commented on pkgs/development/python-modules/manimpango/default.nix:
```diff
@@ -0,0 +1,74 @@
+, buildPythonPackage
+, pkg-config
+, fetchPypi
+, cairo
+, pango
+, pygobject3
+, glib
+, libintl
+, harfbuzz
```
don't much about nix packages but install `pango` should bring in all these dependencies right?
also, manimpango doesn't depend on `pygobject3`.

naveen521kk commented on pkgs/development/python-modules/manimpango/default.nix:
```diff
@@ -0,0 +1,74 @@
+  ];
+
+  buildInputs = [
+    harfbuzz
+    glib
+    cairo
+    libintl
+    pango
```
again, specifying `pango` should get other dependencies right?

naveen521kk commented on pkgs/development/python-modules/manimpango/default.nix:
```diff
@@ -0,0 +1,74 @@
+    pango
+  ];
+
+  propagatedBuildInputs = [
+    pygobject3
+  ];
```
hmm, can you explain what is this?

naveen521kk commented on pkgs/development/python-modules/manimpango/default.nix:
```diff
@@ -0,0 +1,74 @@
+    pygobject3
+  ];
+
+  setupPyBuildFlags = [
+    "-I${cairo.dev}/include/cairo"
+    "-L${cairo.out}/lib"
+    "-lcairo"
+    "-lpangocairo-1.0"
+  ];
```
`setup.py` would get these flags from `pkg-config`, should you specify this manually?

naveen521kk commented on pkgs/development/python-modules/manimpango/default.nix:
```diff
@@ -0,0 +1,74 @@
+    "-lpangocairo-1.0"
+  ];
+
+  # pangocairo is part of normal cairo-package, so remove corresponding deps
+  # from setup.py
+  postPatch =''
+    substituteInPlace setup.py \
+      --replace "\"pangocairo-1.0\"," "" \
+      --replace "_pkg_config = PKG_CONFIG(\"pangocairo\")" "_pkg_config = PKG_CONFIG(\"pango\")"
+  '';
```
can you elaborate on why are you doing this?

naveen521kk commented on pkgs/applications/video/manim-community/default.nix:
```diff
@@ -0,0 +1,141 @@
+
+, cairo
+, ffmpeg_4
+, sox
```
if you are packing `v0.6.0`+, sox isn't required. https://docs.manim.community/en/stable/changelog/0.6.0-changelog.html#breaking-changes

friedelino replied:
ok, done!

naveen521kk commented on pkgs/applications/video/manim-community/default.nix:
```diff
@@ -0,0 +1,141 @@
+  };
+in
+buildPythonApplication rec {
+  pname = "manim-community";
```
hmm, can it be just `manim`? It's so in [AUR](https://aur.archlinux.org/packages/manim/) or [MSYS2](https://packages.msys2.org/base/mingw-w64-manim). But if it can't be then fine.

friedelino replied:
the problem is, we already have a package called `manim` in nixpkgs. unfortunately the existing package is a very outdated version and also not the community developed of manim. Maybe it could be worth considering removing the existing one in favor of manim-community.

mainrs replied:
This is the last change request blocking the PR. How are these votes done? Or can a PR simply remove a package and the final reviewer gives the OK?

collares replied:
cc @JohnAZoidberg

collares commented:
What's the status here? Let us know if help would be useful :)

mofrim committed:
python3Packages.cloup: init at 0.14.0

mofrim committed:
python3Packages.isosurfaces: init at 0.1.0

mainrs commented:
Would you mind updating to 0.13.1?

collares commented:
I think it would be better to merge this PR first and then upgrade to 0.13.1 in a separate PR, to be honest. The reason is that 0.10.0 depends on `skia-pathops`, which is likely to be much harder to package.

mainrs commented:
> I think it would be better to init at <= 0.10.0 first (as soon as the PR builds) and then upgrade to 0.13.1 in a separate PR, to be honest. The reason is that 0.11.0 depends on `skia-pathops` ([ManimCommunity/manim@af462e8](https://github.com/ManimCommunity/manim/commit/af462e83cc3cf8830e8560891a484043ef34ed9b)), which is likely to be much harder to package.

Yep, that's were I failed to build the package. It's such an annoyance tbh... Is there anything that can be done to help here out?

friedelino commented:
Hi there!

I only recently (on christmas ;) continued working on this, trying to provide the most recent version of manim-community... @collares is completely right. I have a local branch for v0.10.0 which at least is building on my system. I will push that later after some cleanup. But the problem with any release > 0.10.0 is the `skia-pathops` library dependency, which doubles the packaging effort for manim. Packaging `skia-pathops` would involve bazel-build-system which is completely new to me and from what I can see in existing packages in nixpkgs, it looks quite complicated.

mainrs commented:
> Hi there!
> 
> I only recently (on christmas ;) continued working on this, trying to provide the most recent version of manim-community... @collares is completely right. I have a local branch for v0.10.0 which at least is building on my system. I will push that later after some cleanup. But the problem with any release > 0.10.0 is the `skia-pathops` library dependency, which doubles the packaging effort for manim. Packaging `skia-pathops` would involve bazel-build-system which is completely new to me and from what I can see in existing packages in nixpkgs, it looks quite complicated.

I've tried to go an easier route and use `pkgs.poetry2nix.mkPoetryApplication` with overwrites to the dependencies that need native stuff. The problem that made it not work for me was that the build somehow didn't find files that were present inside `skia-pathops`. It tried to look for the `pyproject.toml` file but couldn't open it. Checking out the same commit used revealed that the file was actually there.

SuperSandro2000 commented on pkgs/development/python-modules/mapbox-earcut/default.nix:
```diff
@@ -0,0 +1,39 @@
+{ lib
+, buildPythonPackage
+, fetchPypi
+, python3Packages
```
```suggestion
```

SuperSandro2000 commented on pkgs/development/python-modules/mapbox-earcut/default.nix:
```diff
@@ -0,0 +1,39 @@
+    nativeBuildInputs = [ setuptools pybind11 ];
+    propagatedBuildInputs = [ numpy ];
+
+    # Some test failed, some passed...
+    checkInputs = [ pytestCheckHook ];
+    doCheck = false;
```
Can we ignore the failing ones instead?

friedelino replied:
yes -> c04fc24542ff1bac9c1b909df24ea98e3a3fbce7

SuperSandro2000 reviewed (changes requested):
Please add pythonImportsCheck.

friedelino commented on pkgs/development/python-modules/cloup/default.nix:
```diff
@@ -0,0 +1,41 @@
+    longDescription = ''
+      Enriches Click with option groups, constraints, command aliases, help sections for subcommands, themes for --help and other stuff.
+    '';
+    license = licenses.bsd3;
```
c91c98fd02a origin/add-manim-com cloup & isosurfaces: add maintainer

friedelino commented on pkgs/development/python-modules/isosurfaces/default.nix:
```diff
@@ -0,0 +1,32 @@
+      uniform grid, but this uses a quadtree to avoid wasting time sampling
+      many far from the implicit surface.
+    '';
+    license = licenses.mit;
```
c91c98fd02a origin/add-manim-com cloup & isosurfaces: add maintainer

friedelino commented on pkgs/applications/video/manim-community/default.nix:
```diff
@@ -0,0 +1,114 @@
+    pname = "manim";
+    inherit version;
+
+    sha256 = "9005d35a025cc4f9194dcb210cfeeb7e085aa2dfe195d825c39e0391afea9ade";
```
dc9d0a45439 remove empty line from manim-com/default.nix

mofrim reviewed (commented):
unfortunately compilation of openapi-core using python3.9 fails with recent nixpkgs master due to additional failing tests. so i disabled those tests -> 52cce3c8732d9a857ad94a3c1892f4cf96b0e8d4

friedelino commented on pkgs/development/python-modules/mapbox-earcut/default.nix:
```diff
@@ -0,0 +1,39 @@
+, setuptools
+}:
+
+with python3Packages; buildPythonPackage rec {
```
0798ca73cc71c98e1fe0b580a3b2a95289d5e969

friedelino commented on pkgs/development/python-modules/mapbox-earcut/default.nix:
```diff
@@ -0,0 +1,39 @@
+    sha256 = "14z2vj13a1qcaqc6jvg00f367la5470bclvfi9mmlvpac3hhhywd";
+    };
+
+    nativeBuildInputs = [ setuptools pybind11 ];
```
0798ca73cc71c98e1fe0b580a3b2a95289d5e969

SuperSandro2000 reviewed (commented):
Please add pythonImportsChecks.

SuperSandro2000 commented on pkgs/development/python-modules/mapbox-earcut/default.nix:
```diff
@@ -0,0 +1,49 @@
+    sha256 = "14z2vj13a1qcaqc6jvg00f367la5470bclvfi9mmlvpac3hhhywd";
+    };
+
+    buildInputs = [ setuptools pybind11 ];
```
```suggestion
  nativeBuildInputs = [ setuptools ];

    buildInputs = [ pybind11 ];
```

friedelino replied:
6b90e6f626231cbd83843c119cde62f339284599

friedelino commented on pkgs/applications/video/manim-community/default.nix:
```diff
@@ -0,0 +1,115 @@
+    pycairo
+    pydub
+    pygments
+    python
```
53dfd3c1f0f74b21c36b6d58aab109f21c267cf5

friedelino commented on pkgs/development/python-modules/mapbox-earcut/default.nix:
```diff
@@ -0,0 +1,49 @@
+    propagatedBuildInputs = [ numpy ];
+
+    disabledTests = [
+      "test_end_index_too_large"
```
obsolete due to 394926c1663e04d32d6f24a448fe8fb246daded2 and 747e63a9243e3031fb2159f8fca393d590fe7d29

friedelino commented on pkgs/development/python-modules/openapi-core/default.nix:
```diff
@@ -68,6 +68,13 @@ buildPythonPackage rec {
     "test_string_format_invalid_value"
     # Needs a fix for new PyYAML
     "test_django_rest_framework_apiview"
+    "test_response_validator_path_pattern"
```
should be obsolete with 2bfa83550d75f4b0f1c7d52cc29054646d7f6ac2

friedelino commented:
>

friedelino commented:
So, i did some final cleanup... I applied all changes requested by @SuperSandro2000 and @dotlambda. I understand, that it is not a perfect solution to init at a low version in order to avoid more complicated bazel-packaging, but i think merging this can still be a valuable base to start from for future versions.

collares reviewed (approved):
I agree, this should be merged ASAP. Any further improvements can be done in follow-up PRs. Hopefully a committer will show up soon.

mofrim committed:
python3Packages.srt: init at 3.5.2

friedelino commented:
> I agree, this should be merged ASAP. Any further improvements can be done in follow-up PRs. Hopefully a committer will show up soon.

eehhmm... i just took a quick look at all the dependencies and found out that `skia-pathops` was the bazel-pkg i was afraid of. and guess what? it has been packaged! so i was able to init at the most recent version \o/

collares commented:
Result of `nixpkgs-review pr 120479` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>9 packages built:</summary>
  <ul>
    <li>manim-community</li>
    <li>python310Packages.cloup</li>
    <li>python310Packages.isosurfaces</li>
    <li>python310Packages.mapbox-earcut</li>
    <li>python310Packages.srt</li>
    <li>python39Packages.cloup</li>
    <li>python39Packages.isosurfaces</li>
    <li>python39Packages.mapbox-earcut</li>
    <li>python39Packages.srt</li>
  </ul>
</details>

friedelino commented:
Is there anything left i could do to bring this closer to being merged (before something brakes again) or will i just have to wait? @collares

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/508

collares commented:
I truly wish I had commit access to merge this, but I don't :( I've posted on Discourse, but you can also try to ping people on the Matrix channel.

IvarWithoutBones commented on pkgs/development/python-modules/srt/default.nix:
```diff
@@ -0,0 +1,34 @@
+, pythonOlder
+}:
+
+buildPythonPackage rec {
```
Could we use `pythonImportsCheck` here?

friedelino replied:
done!

friedelino replied:
done!

IvarWithoutBones commented on pkgs/development/python-modules/mapbox-earcut/default.nix:
```diff
@@ -0,0 +1,41 @@
+, setuptools
+}:
+
+buildPythonPackage rec {
```
Could we use `pythonImportsCheck` here?

friedelino replied:
done!

friedelino replied:
done!

IvarWithoutBones commented on pkgs/development/python-modules/isosurfaces/default.nix:
```diff
@@ -0,0 +1,33 @@
+, numpy
+}:
+
+buildPythonPackage rec {
```
Could we use `pythonImportsCheck` here?

friedelino replied:
done!

friedelino replied:
of course!

IvarWithoutBones commented on pkgs/development/python-modules/cloup/default.nix:
```diff
@@ -0,0 +1,42 @@
+, pythonOlder
+}:
+
+buildPythonPackage rec {
```
Could we use `pythonImportsCheck` here?

friedelino replied:
of course!

friedelino replied:
done!

IvarWithoutBones commented on pkgs/development/python-modules/isosurfaces/default.nix:
```diff
@@ -0,0 +1,33 @@
+
+  meta = with lib; {
+    homepage = "https://github.com/jared-hughes/isosurfaces";
+    description = "Construct isolines/isosurfaces of a 2D/3D scalar field defined by a function. ";
```
`meta.description` should not end with a period.
```suggestion
    description = "Construct isolines/isosurfaces of a 2D/3D scalar field defined by a function";
```

friedelino replied:
done!

IvarWithoutBones commented on pkgs/development/python-modules/srt/default.nix:
```diff
@@ -0,0 +1,34 @@
+
+  meta = with lib; {
+    homepage = "https://github.com/cdown/srt";
+    description = "Python library for working with SRT files.";
```
```suggestion
    description = "Python library for working with SRT files";
```

IvarWithoutBones commented on pkgs/development/python-modules/cloup/default.nix:
```diff
@@ -0,0 +1,42 @@
+  ];
+
+  buildInputs = [
+    click
```
Shouldn't this be in `propagatedBuildInputs` instead?

friedelino replied:
yeah, right!

IvarWithoutBones reviewed (commented):
Just a few more comments, I'll merge this once those are adressed :+1:

friedelino commented:
> Just a few more comments, I'll merge this once those are adressed +1

alrighty! I've done everything you asked for. If you don't mind @IvarWithoutBones, i have one question for future PRs about practices i am not sure about: if a reviewer asks for a bigger number of changes like you did, is it best practice to make isolated commits for every requested change and keep them unsquashed i.e. visible until you have time to approve them one by one, or is it okay like i did and squash the review-changes to the final set of commits? ah, one more: is it okay or super annoying if i reply to every single change-comment?

IvarWithoutBones commented:
> If you don't mind @IvarWithoutBones, i have one question for future PRs about practices i am not sure about

Of course, I never mind any questions :blush:

> if a reviewer asks for a bigger number of changes like you did, is it best practice to make isolated commits for every requested change and keep them unsquashed i.e. visible until you have time to approve them one by one, or is it okay like i did and squash the review-changes to the final set of commits?

I personally prefer the way you did here, by just squashing the commits to be in the final state they should be. As a reviewer I'd look over all the code again anyways, so if any feedback was wrongfully adressed I can just comment on it based on that. This makes should make the the process of reviewing a bit more streamlined since we can just merge the PR if it looks good, and won't have to wait on a rebase.

>  is it okay or super annoying if i reply to every single change-comment?

I personally do not mind that, although I can see it being a bit annoying if theres a lot of suggestions. I always try to just reply to comments if they're questions, or if I have something to add to it. If every comment would just be "fixed this" I'd recommand to leave a thumbs up emoji and post one comment on the PR itself saying you've adressed all the feedback, so that the reviewer does get a notification for it.

IvarWithoutBones commented on pkgs/development/python-modules/isosurfaces/default.nix:
```diff
@@ -0,0 +1,35 @@
+  };
+
+  buildInputs = [
+    numpy
```
Shouldn't this be in `propagatedBuildInputs` as well?

IvarWithoutBones commented on pkgs/development/python-modules/srt/default.nix:
```diff
@@ -0,0 +1,36 @@
+
+  meta = with lib; {
+    homepage = "https://github.com/cdown/srt";
+    description = "Python library for working with SRT files";
+    longDescription = ''
+      srt is a tiny but featureful Python library for parsing, modifying, and composing SRT files.
+    '';
```
I think we can fit this all in `description`:
```suggestion
    description = "A tiny but featureful Python library for parsing, modifying, and composing SRT files";
```

friedelino replied:
Okay, everything should be fixed now. Thx a lot for your answers to my questions!

IvarWithoutBones replied:
Any time :)

IvarWithoutBones reviewed (changes requested):
Just a few more nits I missed before :)

IvarWithoutBones commented on pkgs/applications/video/manim-community/default.nix:
```diff
@@ -0,0 +1,124 @@
+    tqdm
+    watchdog
+
+    cairo
+    ffmpeg
+    xdg-utils
```
Are these meant to be in `propagatedBuildInputs` and not `buildInputs`?

friedelino replied:
yeah... i've looked into this again and it turned out that we don't need `xdg-utils` at all (i think it was a relict from the existing manim package i started from),  but we do need `ffmpeg`because otherwise manim  cannot produce any video files at runtime => ffmpeg has to be in propagatedBuildInputs. `cairo` is fine in buildInputs.

IvarWithoutBones commented on pkgs/development/python-modules/mapbox-earcut/default.nix:
```diff
@@ -0,0 +1,41 @@
+  src = fetchFromGitHub {
+    owner = "skogler";
+    repo = "mapbox_earcut_python";
+    rev = "v1.0.0";
```
```suggestion
    rev = "v${version}";
```

IvarWithoutBones commented on pkgs/applications/video/manim-community/default.nix:
```diff
@@ -0,0 +1,124 @@
+
+  pythonImportsCheck = [ "manim" ];
+
+  meta = with lib; {
```
I noticed the final binary is called `manim` while `pname` is `manim-community`, that would break `nix run`. Could you add `meta.mainProgram` to fix that?

friedelino replied:
I will wait with that until the naming-issue is settled

IvarWithoutBones reviewed (changes requested):
Looks like I missed a few things again :sweat_smile: 

I also noticed running the `manim` binary from the `nixpkgs-review` shell fails as it cannot import numpy correctly, although it works fine in my regular shell.

naveen521kk commented:
Good to see some progress here. 

A while ago, I asked if the package can be named `manim` instead of `manim-community` as that's how it is on [PyPI](https://pypi.org/projects/manim), [AUR](https://aur.archlinux.org/packages/manim/), [MSYS2](https://packages.msys2.org/base/mingw-w64-manim). See https://github.com/NixOS/nixpkgs/pull/120479#discussion_r641612354. Can that be relooked?

IvarWithoutBones commented:
> Good to see some progress here.
> 
> A while ago, I asked if the package can be named `manim` instead of `manim-community` as that's how it is on [PyPI](https://pypi.org/projects/manim), [AUR](https://aur.archlinux.org/packages/manim/), [MSYS2](https://packages.msys2.org/base/mingw-w64-manim). See [#120479 (comment)](https://github.com/NixOS/nixpkgs/pull/120479#discussion_r641612354). Can that be relooked?

I think the issue is that `manim` already exists in nixpkgs, see repology: https://repology.org/project/manim/versions

collares commented:
Note that the AUR version has `pkgbase = manimce` and `pkgname = manim`, so it shows up under `manimce`on Repology: https://repology.org/project/manimce/versions

naveen521kk commented:
Just to clear some confusion:
- https://github.com/3b1b/manim is called `manimgl` on PyPI: https://pypi.org/project/manimgl
- https://github.com/ManimCommunity/manim is called `manim` on PyPI: https://pypi.org/project/manim

The version that I think is currently present in nixpkgs is from the `manimlib` python package, which is unmaintained, see https://github.com/3b1b/manim/issues/1243 and https://docs.manim.community/en/stable/installation/versions.html.
(I inferred that from https://github.com/NixOS/nixpkgs/blob/dfd82985c273aac6eced03625f454b334daae2e8/pkgs/applications/video/manim/default.nix#L10-L13 but not sure). I think replacing that with the community version is better; a new package for `manimgl` should do for the 3b1b's version (if someone is interested in maintaining it).

On the other hand, if this package is not going to be called `manim`, I would like it to be called `manimce`, as that's how it is on [Chocolatey](https://community.chocolatey.org/packages/manimce) and other places.

collares commented:
Iindeed, I didn't know about the `manimlib`/`manimgl` distinction. Thanks for clarifying!

friedelino commented:
Thanks @naveen521kk for your research about the different manim's. If we don't remove the existing outdated and unmaintained  manim-package from nixpkgs the package name i would prefer is "manimce". How  can we  decide on removing the existing one? I am curious: Do we need some kind of poll for removing something from nixpkgs?

collares commented:
cc @JohnAZoidberg (`manim` maintainer) and @hrhino (who fixed the package last time it broke): Do you still use `manimlib`? Upstream is now developing `manimgl`, as @naveen521kk explained above. Would you be OK with removing `manimlib` in favor of `ManimCommunity`, and perhaps adding `manimgl` in the future?

collares commented:
I think the safest way to proceed is to change the current commit to rename the package to `manimce`, and then add a separate commit that deletes the old `manim` and renames `manimce` to `manim`. Then this should be ready to merge, and one of the commits can be easily reverted separately if the maintainer disagrees about the decision. Typically I'd say we should wait a bit for a reply, but the ping in https://github.com/NixOS/nixpkgs/pull/120479#discussion_r787755127 happened four months ago and this PR has been in queue for a long time.

friedelino commented:
> I think the safest way to proceed is to change the current commit to rename the package to `manimce`, and then add a separate commit that deletes the old `manim` and renames `manimce` to `manim`. Then this should be ready to merge...

Sounds like a plan @collares! Unfortunately `nixpkgs-review pr 120479` now fails with some bizarre errors:

```
> nixpkgs-review pr 120479
$ git -c fetch.prune=false fetch --no-tags --force https://github.com/NixOS/nixpkgs master:refs/nixpkgs-review/0 pull/120479/head:refs/nixpkgs-review/1
$ git worktree add /home/.../.cache/nixpkgs-review/pr-120479-18/nixpkgs a0b95bb245ac5f6da250d6b6b12b596f3c6fd6df
Bereite Arbeitsverzeichnis vor (losgelöster HEAD a0b95bb245a)
Aktualisiere Dateien: 100% (30603/30603), fertig.
HEAD ist jetzt bei a0b95bb245a Merge pull request #174168 from fabaff/hahomematic-bump
$ nix-env --option system x86_64-linux -f /home/.../.cache/nixpkgs-review/pr-120479-18/nixpkgs -qaP --xml --out-path --show-trace
error: undefined variable 'TestMore'

       at /home/.../.cache/nixpkgs-review/pr-120479-18/nixpkgs/pkgs/top-level/perl-packages.nix:16951:87:

        16950|     };
        16951|     buildInputs = [ DevelHide FileShareDirInstall ParallelForkManager TestDifferences TestMore TestPod TestPodCoverage TestWarnings ];
             |                                                                                       ^
        16952|     propagatedBuildInputs = [ DataDumperConcise DBMDeep DevelCheckOS FileFindRule FileShareDir ];
....
....
       at /home/..../.cache/nixpkgs-review/pr-120479-18/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:278:7:

          277|     // (lib.optionalAttrs (attrs ? name || (attrs ? pname && attrs ? version)) {
          278|       name =
             |       ^
          279|         let

       … while querying the derivation named 'perl5.32.1-Number-Phone-3.8004'
https://github.com/NixOS/nixpkgs/pull/120479 failed to build
$ git worktree prune

```

but it looks like it doesn't have to do anything with my recent commits, i hope....

collares commented:
#174193

dotlambda commented:
The attribute name in `python-packages.nix` and the `pname` should match what it's called on PyPI.

dotlambda commented:
`manimlib` can be removed and either a `throw` or an actual alias added to `python-aliases.nix`. Preferably the former if the difference in API is big.

mofrim committed:
manim: remove to be replaced by ManimCE

Existing manim package has been incorrectly named manim, although on PyPi it is
called manimlib. It is now developed under the name manimgl, so the PyPi
package manimlib will receive no further updates. The new manim
package in nixpkgs is now the Manim Community Edition.

mofrim committed:
manim: init at 0.15.2

friedelino commented:
Alrighty. I have now added a commit, which removes old `manim` pkg, called `manim: rename to manimlib and remove` because the alias I put into `top-level/aliases.nix` has to be for something else but `manim`. Also I renamed the new package accordingly. I wondered if I should create an extra PR for the manim(lib) removal, though. I guess you will let me know :wink:

dotlambda commented:
I didn't realize the old package was called `manim`, my bad. No need for an entry in `aliases.nix` in that case.

friedelino commented:
> I didn't realize the old package was called `manim`, my bad. No need for an entry in `aliases.nix` in that case.

Aha! Alright, changed the commit. Now the old `manim` just gets replaced by the community one... :crossed_fingers:

friedelino commented:
There still is the issue remaining that you cannot run `results/bin/manim` in nixpkgs-review shell. It fails with:

```
foo@bar .cache/nixpkgs-review/pr-120479-23 via ❄ impure shell (review-shell)
$ manim
Traceback (most recent call last):
  File "/nix/store/i3prvnmnqz7znxmm46vfbq1zggzjqdd3-python3.10-numpy-1.21.5/lib/python3.10/site-packages/numpy/core/__init__.py", line 22, in <module>
    from . import multiarray
  File "/nix/store/i3prvnmnqz7znxmm46vfbq1zggzjqdd3-python3.10-numpy-1.21.5/lib/python3.10/site-packages/numpy/core/multiarray.py", line 12, in <module>
    from . import overrides
  File "/nix/store/i3prvnmnqz7znxmm46vfbq1zggzjqdd3-python3.10-numpy-1.21.5/lib/python3.10/site-packages/numpy/core/overrides.py", line 7, in <module>
    from numpy.core._multiarray_umath import (
ModuleNotFoundError: No module named 'numpy.core._multiarray_umath'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/nix/store/31bafgcaplh5jdm3w273pq09s5vkmqhn-manim-0.15.2/bin/.manim-wrapped", line 6, in <module>
    from manim.__main__ import main
  File "/nix/store/31bafgcaplh5jdm3w273pq09s5vkmqhn-manim-0.15.2/lib/python3.9/site-packages/manim/__init__.py", line 15, in <module>
    from ._config import *
  File "/nix/store/31bafgcaplh5jdm3w273pq09s5vkmqhn-manim-0.15.2/lib/python3.9/site-packages/manim/_config/__init__.py", line 10, in <module>
    from .utils import ManimConfig, ManimFrame, make_config_parser
  File "/nix/store/31bafgcaplh5jdm3w273pq09s5vkmqhn-manim-0.15.2/lib/python3.9/site-packages/manim/_config/utils.py", line 27, in <module>
    import numpy as np
  File "/nix/store/i3prvnmnqz7znxmm46vfbq1zggzjqdd3-python3.10-numpy-1.21.5/lib/python3.10/site-packages/numpy/__init__.py", line 150, in <module>
    from . import core
  File "/nix/store/i3prvnmnqz7znxmm46vfbq1zggzjqdd3-python3.10-numpy-1.21.5/lib/python3.10/site-packages/numpy/core/__init__.py", line 48, in <module>
    raise ImportError(msg)
ImportError:

IMPORTANT: PLEASE READ THIS FOR ADVICE ON HOW TO SOLVE THIS ISSUE!

Importing the numpy C-extensions failed. This error can happen for
many reasons, often due to issues with your setup or how NumPy was
installed.

We have compiled some common reasons and troubleshooting tips at:

    https://numpy.org/devdocs/user/troubleshooting-importerror.html

Please note and check the following:

  * The Python version is: Python3.9 from "/nix/store/vzqny68wq33dcg4hkdala51n5vqhpnwc-python3-3.9.12/bin/python3.9"
  * The NumPy version is: "1.21.5"

and make sure that they are the versions you expect.
Please carefully study the documentation linked above for further help.

Original error was: No module named 'numpy.core._multiarray_umath'
```
On the other hand it runs happily when you build it normally... i don't understand how the first error message comes from python3.10 but the latter one from python3.9.

Could it be possible that it is a problem that we have both python3.10 and python3.9 versions of Packages available and they somehow get mixed up? Or that the nixpkgs-review-shell? also checked the wrapped "binaries" in `results/manim/bin`. They are solely wrapped around python3.9. So i don't understand where the python3.10 in the first error message comes from.

collares commented:
I think you're right, the nixpkgs-review shell has both Python 3.9 and 3.10 and that causes problems. https://github.com/NixOS/nixpkgs/pull/150972#issuecomment-1029543308 provides a good description of the problem.

friedelino commented:
> I think you're right, the nixpkgs-review shell has both Python 3.9 and 3.10 and that causes problems. [#150972 (comment)](https://github.com/NixOS/nixpkgs/pull/150972#issuecomment-1029543308) provides a good description of the problem. You can test the manim binary by passing `-p manim` to nixpkgs-review.

i guess you mean passing `-p manim` to **nix-shell** not `nixpkgs-review` ?! *my* nixpkgs-review does not have a -p option. but anyhow, this i already knew and did before. with `nix-shell -p manim` and after `nix-build` in my nixpkgs-repo everything works fine.

collares commented:
Sorry, I should have typed the complete command! I meant `nixpkgs-review pr 120479 -p manim`.

dotlambda commented on pkgs/applications/video/manim/default.nix:
```diff
@@ -1,64 +1,123 @@
+    fundus-calligra microtype wasysym physics dvisvgm jknapltx wasy cm-super
+    babel-english gnu-freefont mathastext cbfonts-fd;
+  };
+in
+with python3.pkgs; buildPythonApplication rec {
```
```suggestion
in python3.pkgs.buildPythonApplication rec {
```

dotlambda commented on pkgs/applications/video/manim/default.nix:
```diff
@@ -1,64 +1,123 @@
   };
 
-  patches = [ ./remove-dependency-constraints.patch ];
+  nativeBuildInputs = [ poetry-core ];
```
```suggestion
  nativeBuildInputs = [
    python3.pkgs.poetry-core
  ];
```

dotlambda commented on pkgs/applications/video/manim/default.nix:
```diff
@@ -1,64 +1,123 @@
-  nativeBuildInputs = [ pbr ];
+  buildInputs = [ cairo ];
 
   propagatedBuildInputs = [
```
```suggestion
  propagatedBuildInputs = with python3.pkgs; [
```

dotlambda commented on pkgs/applications/video/manim/default.nix:
```diff
@@ -1,64 +1,123 @@
+    watchdog
 
-    cairo sox ffmpeg xdg-utils
+    ffmpeg
+
+    (texlive.combine manim-tinytex)
```
Only Python packages belong in `propagatedBuildInputs`.
```suggestion
```
You could use `makeWrapperArgs` instead.

dotlambda commented on pkgs/applications/video/manim/default.nix:
```diff
@@ -1,64 +1,123 @@
-        | tee | grep -F "ISO Media, MP4 Base Media v1 [ISO 14496-12:2003]"
-    done
-  '';
+  doCheck = true;
```
```suggestion
```
This is automatically set when using `buildPythonApplication`.
However please specify `checkPhase` or use `pytestCheckHook`.

dotlambda reviewed (changes requested):
Most commit messages need to be prefixed with `python3Packages.`.
The first and last commits should be squashed together.
Isosurfaces needs to set `doCheck = false` with a comment that upstream has no tests.

collares commented:
It was my suggestion to not squash the first and last commits. I don't really care either way, though, whatever gets this merged faster.

dotlambda commented:
> It was my suggestion to not squash the first and last commits. I don't really care either way, though; whatever gets this merged faster is fine by me.

Fine with me if it was a concious decision :+1:

collares commented on pkgs/applications/video/manim/default.nix:
```diff
@@ -1,64 +1,123 @@
+    watchdog
 
-    cairo sox ffmpeg xdg-utils
+    ffmpeg
+
+    (texlive.combine manim-tinytex)
   ];
```
To speed things up, I think @dotlambda's suggestion is:

```suggestion
  ];

  makeWrapperArgs = [ "--prefix PATH : ${lib.makeBinPath [ ffmpeg (texlive.combine manim-tinytex) ]}" ];
```

I haven't tested it, though.

friedelino commented:
so... i committed all suggestions. about now at least partially enabled tests for manim: there might be still work to do if we want to get all tests running. especially most failing tests need display. I tried to do it with `xvfb` but it didn't work. btw: are there any updates on the 'tests need display'-problem for general package development, like a go-to-way or best-practice ?

dotlambda reviewed (approved):
I think it's fine to not have all tests work for now. Let's hope some day someone comes up with a way to fix them.

SuperSandro2000 reviewed (commented):
With 184 comments this is well looked at. Now get it out there and we find the remaining bugs there.

dotlambda commented on pkgs/applications/video/manim/default.nix:
```diff
@@ -1,64 +1,134 @@
 
-  disabled = pythonOlder "3.7";
+  checkInputs = [
+    python3.pkgs.pytest-cov
```
What @SuperSandro2000 means is that we're not interested in testing coverage, so this dependency should be patched out, e.g. with the next update.

SuperSandro2000 replied:
Already created a PR that fixes that

friedelino commented:
Woohoo! Finally, it is merged :partying_face: :partying_face: :partying_face:

Thanks to everyone who helped me with your comments & patience! I think i learned quite a bit about nix packaging :)
